File: api/rest/article.go
package rest

import (
	"context"
	"net/http"
	"time"

	"github.com/labstack/echo/v4"
	"github.com/ryutah/realworld-echo/api/rest/gen"
	"github.com/ryutah/realworld-echo/usecase"
)

type getArticleOutputPort struct{}

func NewGetArticleOutputPort(e usecase.ErrorOutputPort) usecase.OKOutputPort[usecase.GetArticleResult] {
	return &getArticleOutputPort{}
}

func (g *getArticleOutputPort) OK(ctx context.Context, _ usecase.GetArticleResult) error {
	c := echoContextFromContext(ctx)
	return c.JSON(http.StatusOK, gen.SingleArticleResponse{
		Article: gen.Article{
			Author: gen.Profile{
				Bio:       "dummy",
				Following: false,
				Image:     "dummy",
				Username:  "dummy",
			},
			Body:           "dummy",
			CreatedAt:      time.Now(),
			Description:    "dummy",
			Favorited:      false,
			FavoritesCount: 0,
			Slug:           "dummy",
			TagList:        []string{},
			Title:          "",
			UpdatedAt:      time.Now(),
		},
	})
}

type Artcle struct {
	inputPort struct {
		getArticle usecase.GetArticleInputPort
	}
}

func NewArticle(getArticle usecase.GetArticleInputPort) *Artcle {
	return &Artcle{
		inputPort: struct {
			getArticle usecase.GetArticleInputPort
		}{
			getArticle: getArticle,
		},
	}
}

func (a *Artcle) GetArticle(c echo.Context, slug string) error {
	ctx := newContext(c)
	return a.inputPort.getArticle.Get(ctx, slug)
}

File: realworld-web/app/domain/article.ts
import { Profile } from "@/app/domain/profile";
import { NonFunctionProperties } from "@/app/lib/types";

export type ArticleProps = NonFunctionProperties<Article>;

export class Article {
  readonly slug: string;
  readonly title: string;
  readonly description: string;
  readonly body: string;
  readonly tagList: string[];
  readonly createdAt: Date;
  readonly updatedAt: Date;
  readonly favoritesCount: number;
  readonly favorited?: boolean;
  readonly author: Profile;

  constructor(article: ArticleProps) {
    this.slug = article.slug;
    this.title = article.title;
    this.description = article.description;
    this.body = article.body;
    this.tagList = article.tagList;
    this.createdAt = article.createdAt;
    this.updatedAt = article.updatedAt;
    this.favoritesCount = article.favoritesCount;
    this.author = article.author;
  }

  public formattedCreatedAt(): string {
    return this.createdAt.toDateString();
  }

  public formattedUpdatedAt(): string {
    return this.updatedAt.toDateString();
  }
}
